src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,907 @@
+  | GetAddressPayload
+  | SignTransactionPayload
+  | SignMessagePayload
+  | ConnectionChangedPayload
+);
+
+interface IFrameMessageResponse {
+  success: boolean;
+  action: IFrameMessageAction;
+  messageId: number;
+  payload: IFrameMessageResponsePayload;
+  error?: unknown;
+}
+
+export interface AccountDetails {
+  index?: number;
+  bip44?: boolean;
+  hdPath?: string;
+}
+
+export interface LedgerBridgeKeyringOptions {
+  hdPath?: string;
+  accounts?: string[];
+  accountDetails?: Record<string, AccountDetails>;
+  accountIndexes?: Record<string, number>;
+  bridgeUrl?: string;
+  implementFullBIP44?: boolean;
+}
+
+/**
+ * Check if the given transaction is made with ethereumjs-tx or @ethereumjs/tx
+ *
+ * Transactions built with older versions of ethereumjs-tx have a
+ * getChainId method that newer versions do not.
+ * Older versions are mutable
+ * while newer versions default to being immutable.
+ * Expected shape and type
+ * of data for v, r and s differ (Buffer (old) vs BN (new)).
+ *
+ * @param tx
+ * @returns Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(
+  tx: TypedTransaction | OldEthJsTransaction,
+): tx is OldEthJsTransaction {
+  return typeof (tx as OldEthJsTransaction).getChainId === 'function';
+}
+
+function isSignTransactionResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is SignTransactionPayload {
+  return typeof (payload as SignTransactionPayload).v === 'string';
+}
+
+function isSignMessageResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is SignMessagePayload {
+  return typeof (payload as SignMessagePayload).v === 'number';
+}
+
+function isGetAddressMessageResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is GetAddressPayload {
+  return typeof (payload as GetAddressPayload).publicKey === 'string';
+}
+
+function isConnectionChangedResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is ConnectionChangedPayload {
+  return typeof (payload as ConnectionChangedPayload).connected === 'boolean';
+}
+
+export class LedgerBridgeKeyring extends EventEmitter {
+  static type: string = keyringType;
+
+  readonly type: string = keyringType;
+
+  page = 0;
+
+  perPage = 5;
+
+  unlockedAccount = 0;
+
+  accounts: readonly string[] = [];
+
+  accountDetails: Record<string, AccountDetails> = {};
+
+  hdk = new HDKey();
+
+  hdPath = hdPathString;
+
+  paths: Record<string, number> = {};
+
+  network: keyof typeof NETWORK_API_URLS = 'mainnet';
+
+  implementFullBIP44 = false;
+
+  iframeLoaded = false;
+
+  isDeviceConnected = false;
+
+  currentMessageId = 0;
+
+  messageCallbacks: Record<number, (response: IFrameMessageResponse) => void> =
+    {};
+
+  bridgeUrl: string = BRIDGE_URL;
+
+  iframe?: HTMLIFrameElement;
+
+  delayedPromise?: {
+    resolve: (value: unknown) => void;
+    reject: (error: unknown) => void;
+    transportType: string;
+  };
+
+  private _eventListener?: (params: {
```

Comment:
We should use the sharp notation here as for private methods, but we should then find an alternative to the current tests as they stub this variable

Reply:
we can't spy on `#eventListener` directly, but we can make sure that `global.window.removeEventListener` has been called with two arguments: "message" and a function.

Reply:
we can potentially transform `#eventListener` into a method instead of a variable, as it seems like we never unset it after calling `#setupListener()` in the constructor. 

@Gudahtt what do you think? do you know if there is a reason why we want to set it like this?

Reply:
Making this a method would make sense to me; I'm not aware of any reason for it to be set dynamically as a property